### PR TITLE
Explain ## identifiers in manual

### PR DESCRIPTION
**Pull request by lassik**
Lightly edited from [Marc's comment](https://github.com/gambit/gambit/commit/e05d171759c9453d5529021818ee4774afb5412c).

### CODE REVIEW

File: doc/gambit.txi [hunk 1]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+
+@example
+(define (foo x)
+    (if (##vector? x)
+        (##vector-ref x 5)
+        (##fail-check-vector '(1 . x) foo x)))
+@end example
+
+If the code is in the scope of a @code{(declare-safe-define-procedure
+#t)} then it is vector-ref that is called instead of @code{##vector-ref}
```
Comment (feeley) [inline doc/gambit.txi:351]: Please add `@code{vector-ref}`.

File: doc/gambit.txi [hunk 2]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+(##println (foo (bar 0.5)))
+(##pp#unimplemented "hello")
+@end example
+
+The namespace declarations in @code{~~lib/gambit/prim/prim#.scm} have
+caused a mapping of square to @code{##square}, car to @code{##car} and
+println to @code{##println} because those primitives perform the same
+operations (assuming the code has no errors). Note that foo and bar have
+remained the same (because they are not procedures exported by the
+runtime library) but pp has been mapped to @code{##pp#unimplemented}
```
Comment (feeley) [inline doc/gambit.txi:304]: Please add `@code{pp}`, also on next line.

File: doc/gambit.txi [hunk 3]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+@example
+(##define (foo x) (##square (##car x)))
+(##println (foo (bar 0.5)))
+(##pp#unimplemented "hello")
+@end example
+
+The namespace declarations in @code{~~lib/gambit/prim/prim#.scm} have
+caused a mapping of square to @code{##square}, car to @code{##car} and
+println to @code{##println} because those primitives perform the same
+operations (assuming the code has no errors). Note that foo and bar have
```
Comment (feeley) [inline doc/gambit.txi:302]: Please add `@code{foo}` and `@code{bar}`.

File: doc/gambit.txi [hunk 4]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+is equivalent to this code:
+
+@example
+(##define (foo x) (##square (##car x)))
+(##println (foo (bar 0.5)))
+(##pp#unimplemented "hello")
+@end example
+
+The namespace declarations in @code{~~lib/gambit/prim/prim#.scm} have
+caused a mapping of square to @code{##square}, car to @code{##car} and
```
Comment (feeley) [inline doc/gambit.txi:300]: Please add `@code{square}` and `@code{car}`.

File: doc/gambit.txi [hunk 5]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+prefix. The purpose of the @code{primitive} special form is to abstract
+from the @code{##} prefix. The call @code{(primitive foo)} is equivalent
+to @code{##foo} and @code{(primitive (foo a b c))} is equivalent to
+@code{(##foo a b c)}. Another Scheme implementation could define
+@code{primitive} differently to use a different prefix (or no prefix) or
+implement an arbitrary mapping to implementation dependent identifiers.
+
+The @code{primitive} special form works in tandem with the
+@code{define-procedure} and @code{define-primitive} forms and the
+@code{~~lib/gambit/prim/prim#.scm} file and @code{gambit/prim} library.
```
Comment (feeley) [inline doc/gambit.txi:276]: Probably best to use the library name `(gambit prim)` to conform to R7RS.  Also a few lines below.

File: doc/gambit.txi [hunk 6]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+@end example
+
+The namespace declarations in @code{~~lib/gambit/prim/prim#.scm} have
+caused a mapping of square to @code{##square}, car to @code{##car} and
+println to @code{##println} because those primitives perform the same
+operations (assuming the code has no errors). Note that foo and bar have
+remained the same (because they are not procedures exported by the
+runtime library) but pp has been mapped to @code{##pp#unimplemented}
+because pp is a procedure exported by the runtime library but
+@code{##pp} is not defined. Having unimplemented in the name will help
```
Comment (feeley) [inline doc/gambit.txi:306]: Use `@code{unimplemented}`.

File: doc/gambit.txi [hunk 7]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+catch situations where the programmer expected a primitive operation to
+exist but this isn't the case.
+
+The @code{define-procedure} form does two things. It supports type
+annotations in the parameter list and it inserts a @code{(include
+"~~lib/gambit/prim/prim#.scm")} in the body so that primitive operations
+can be used without the @code{##} prefix. Type checking and automatic
+forcing of promise arguments is also added implicitly. The form
```
Comment (feeley) [inline doc/gambit.txi:314]: is -> are

File: doc/gambit.txi [hunk 8]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+@code{define-primitive} is similar, but the procedure defined is
+implicitly prefixed with @code{##}.
+
+So all of these things work together to abstract away from the concept
+of "primitive" operations. "Primitives" are implemented using procedures
+with a @code{##} prefix, but other Scheme implementations could do it
+differently.
+
+Finally, there's the @code{(declare-safe-define-procedure <bool>)} that
```
Comment (feeley) [inline doc/gambit.txi:323]: that -> form that

File: doc/gambit.txi [hunk 9]
```diff
@@ -149,6 +149,7 @@ repository.
 
 @menu
 * Accessing the system files::      Accessing the system files
+* Extending the core system::       Extending the core system
```
Comment (feeley) [inline doc/gambit.txi:152]: This is too early in the documentation.  It is related to modules so should go in that chapter.

File: doc/gambit.txi [hunk 10]
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
...
+@example
+(define (foo x)
+    (if (##vector? x)
+        (##vector-ref x 5)
+        (##fail-check-vector '(1 . x) foo x)))
+@end example
+
+If the code is in the scope of a @code{(declare-safe-define-procedure
+#t)} then it is vector-ref that is called instead of @code{##vector-ref}
+which will both check that x is a vector (redundantly) and that the
```
Comment (feeley) [inline doc/gambit.txi:352]: x - > `@code{x}`

**Comment by feeley** [comment]
It would be good to add this table of type specifiers to the description... https://github.com/gambit/gambit/pull/541#issuecomment-622531223

**Comment by lassik** [comment]
OK, fixed all of those. Added table and moved to Modules chapter as "Primitive Procedures" (maybe not the best name, but it's something to start with).

**Pull request merged**